{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "absolute-preview",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#KEYS\" data-toc-modified-id=\"KEYS-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>KEYS</a></span></li><li><span><a href=\"#Prepping-read-count-data\" data-toc-modified-id=\"Prepping-read-count-data-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Prepping read count data</a></span><ul class=\"toc-item\"><li><span><a href=\"#Combining-unnormalized-raw-featurecounts\" data-toc-modified-id=\"Combining-unnormalized-raw-featurecounts-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Combining unnormalized raw featurecounts</a></span></li><li><span><a href=\"#Gathering-fragment-lengths-from-Featurecounts\" data-toc-modified-id=\"Gathering-fragment-lengths-from-Featurecounts-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Gathering fragment lengths from Featurecounts</a></span></li></ul></li><li><span><a href=\"#Installing-packages-and-loading-libraries\" data-toc-modified-id=\"Installing-packages-and-loading-libraries-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Installing packages and loading libraries</a></span></li><li><span><a href=\"#Size-Factors\" data-toc-modified-id=\"Size-Factors-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Size Factors</a></span><ul class=\"toc-item\"><li><span><a href=\"#DESeq2-normalizing-raw-data\" data-toc-modified-id=\"DESeq2-normalizing-raw-data-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>DESeq2 normalizing raw data</a></span></li><li><span><a href=\"#Adding-GC-content-to-DESeq2-normalized-gene-counts\" data-toc-modified-id=\"Adding-GC-content-to-DESeq2-normalized-gene-counts-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Adding GC content to DESeq2 normalized gene counts</a></span></li></ul></li><li><span><a href=\"#DESeq2---Differential-Expression-Analysis\" data-toc-modified-id=\"DESeq2---Differential-Expression-Analysis-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>DESeq2 - Differential Expression Analysis</a></span><ul class=\"toc-item\"><li><span><a href=\"#CC-2931\" data-toc-modified-id=\"CC-2931-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>CC-2931</a></span></li><li><span><a href=\"#CC-2344\" data-toc-modified-id=\"CC-2344-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>CC-2344</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "common-tuition",
   "metadata": {},
   "source": [
    "## KEYS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "macro-ranking",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy import random\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib.figure import Figure\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "from scipy import stats\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import fligner\n",
    "from scipy.stats import skewtest\n",
    "stats.junk = lambda chisq, df: stats.chi2.sf(chisq, df)\n",
    "import csv\n",
    "import gffpandas.gffpandas as gffpd\n",
    "import sklearn\n",
    "from sklearn.metrics import r2_score\n",
    "import os\n",
    "import statistics\n",
    "import statsmodels.formula.api as smf\n",
    "from statsmodels.formula.api import ols\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "CC2344 = [\"CC2344-L1\", \"CC2344-L2\", \"CC2344-L3\", \"CC2344-L4\", \"CC2344-L5\", \"CC2344-L6\", \"CC2344-L7\", \"CC2344-L8\", \"CC2344-L9\", \"CC2344-L10\", \"CC2344-L11\", \"CC2344-L12\", \"CC2344-L13\", \"CC2344-L14\", \"CC2344-L15\"]\n",
    "CC2931 = [\"CC2931-L1\", \"CC2931-L2\", \"CC2931-L3\", \"CC2931-L4\", \"CC2931-L5\", \"CC2931-L6\", \"CC2931-L7\", \"CC2931-L9\", \"CC2931-L10\", \"CC2931-L11\", \"CC2931-L13\", \"CC2931-L14\", \"CC2931-L15\"]\n",
    "\n",
    "#### Dataframe recording the generation time per sample ####\n",
    "dic_gen = {}\n",
    "generation = pd.read_csv('/research/projects/chlamydomonas/MAexpression/genome_info/mutation_info/Mutation_Fitness.txt', delimiter = '\\t')\n",
    "generation['Sample'] = generation['Sample'].str.replace('_', '-L')\n",
    "generation = generation.loc[generation['Sample'].isin(CC2344 + CC2931)]\n",
    "\n",
    "for i in generation.index.values:\n",
    "    dic_gen[generation.at[i,'Sample']] = generation.at[i, 'Generation']\n",
    "generations = pd.Series(dic_gen)\n",
    "\n",
    "#### Dataframe recording the number of mutations per sample ####\n",
    "all_mutations = pd.read_csv('/research/projects/chlamydomonas/MAexpression/genome_info/mutation_info/all_mutations.csv', delimiter = '\\t')\n",
    "dic_mut = {maline:all_mutations.loc[all_mutations['sample'] == maline].count()[0] for maline in all_mutations['sample'].values.tolist()}\n",
    "mutations = pd.Series(dic_mut)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "provincial-begin",
   "metadata": {},
   "source": [
    "## Prepping read count data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "successful-myanmar",
   "metadata": {},
   "source": [
    "### Combining unnormalized raw featurecounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "periodic-westminster",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################\n",
    "#### CC-2344 ####\n",
    "#################\n",
    "\n",
    "#### NAMES OF GENE COUNT FILES ####\n",
    "keyword = 'genes_count'\n",
    "label = []\n",
    "for fname in os.listdir('/research/projects/chlamydomonas/MAexpression/data/gene_count/'):\n",
    "    if keyword in fname:\n",
    "        fname = fname.replace('_genes_count','')\n",
    "        if 'CC2344' in fname:\n",
    "            label.append(fname)\n",
    "            \n",
    "#### CC2344 - UNNORMALIZED GENE COUNTS ####\n",
    "gene_count = pd.DataFrame()\n",
    "for i in label:\n",
    "    section = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/gene_count/' + i + '_genes_count', delimiter = '\\t', skiprows = 1)\n",
    "    section.columns = ['geneid', 'chr', 'start', 'end', 'strand', 'length', i]\n",
    "    section = section[['geneid', i]]\n",
    "    gene_count = pd.concat([gene_count, section], axis = 1)\n",
    "gene_count = gene_count.loc[:,~gene_count.columns.duplicated()]\n",
    "gene_count = gene_count.set_index('geneid')\n",
    "gene_count.columns = pd.Series(gene_count.columns).str.split('.', expand = True)[1]\n",
    "\n",
    "#### EXPORTING ####\n",
    "gene_count.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2344_raw.txt', sep = '\\t', index = True, header = True)\n",
    "\n",
    "#################\n",
    "#### CC-2931 ####\n",
    "#################\n",
    "\n",
    "#### NAMES OF GENE COUNT FILES ####\n",
    "keyword = 'genes_count'\n",
    "label = []\n",
    "for fname in os.listdir('/research/projects/chlamydomonas/MAexpression/data/gene_count/'):\n",
    "    if keyword in fname:\n",
    "        fname = fname.replace('_genes_count','')\n",
    "        if 'CC2931' in fname:\n",
    "            label.append(fname)\n",
    "            \n",
    "#### CC2931 - UNNORMALIZED GENE COUNTS ####\n",
    "gene_count = pd.DataFrame()\n",
    "for i in label:\n",
    "    section = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/gene_count/' + i + '_genes_count', delimiter = '\\t', skiprows = 1)\n",
    "    section.columns = ['geneid', 'chr', 'start', 'end', 'strand', 'length', i]\n",
    "    section = section[['geneid', i]]\n",
    "    gene_count = pd.concat([gene_count, section], axis = 1)\n",
    "gene_count = gene_count.loc[:,~gene_count.columns.duplicated()]\n",
    "gene_count = gene_count.set_index('geneid')\n",
    "gene_count.columns = pd.Series(gene_count.columns).str.split('.', expand = True)[1]\n",
    "\n",
    "#### EXPORTING ####\n",
    "gene_count.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2931_raw.txt', sep = '\\t', index = True, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "offensive-morocco",
   "metadata": {},
   "source": [
    "### Gathering fragment lengths from Featurecounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "charged-particle",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragment_length = pd.read_csv('/research/projects/chlamydomonas/MAexpression/data/gene_count/198.CC2344-L15-rep3_genes_count', delimiter = '\\t', skiprows = 1, usecols = ['Geneid','Length'])\n",
    "fragment_length = fragment_length.set_index('Geneid')\n",
    "\n",
    "# EXPORT DATA\n",
    "fragment_length.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/fpkm/fragment_length.csv', sep = '\\t', index = True, index_label= 'index', header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "metallic-signal",
   "metadata": {},
   "source": [
    "## Installing packages and loading libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "confidential-serum",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: ── \u001b[1mAttaching packages\u001b[22m ───────────────────────────────── tidyverse 1.3.1.\u001b[31m9000\u001b[39m ──\n",
      "\n",
      "R[write to console]: \u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.4.1     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 1.0.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.2.0     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.1.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.3.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.0.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "R[write to console]: ── \u001b[1mConflicts\u001b[22m ───────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: MatrixGenerics\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "## TIDYVERSE\n",
    "library(tidyverse)\n",
    "\n",
    "## DESEQ2\n",
    "library(\"BiocParallel\")\n",
    "register(MulticoreParam(4))\n",
    "library(\"DESeq2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "catholic-finger",
   "metadata": {},
   "source": [
    "## Size Factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "necessary-going",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CC2931.ANC.rep1 CC2931.ANC.rep2 CC2931.ANC.rep3  CC2931.L1.rep1  CC2931.L1.rep2 \n",
      "      1.8502940       1.0322654       0.8007557       2.3773915       1.3148481 \n",
      " CC2931.L1.rep3  CC2931.L2.rep1  CC2931.L2.rep2  CC2931.L2.rep3 CC2931.L15_rep1 \n",
      "      1.4059812       0.5266999       0.9162189       1.0550166       1.4258836 \n",
      "CC2931.L15_rep2 CC2931.L15_rep3  CC2931.L3_rep1  CC2931.L3_rep2  CC2931.L3_rep3 \n",
      "      1.0288307       1.2657351       0.6031843       1.1181484       0.8886967 \n",
      " CC2931.L4_rep1  CC2931.L4_rep2  CC2931.L5_rep1  CC2931.L5_rep3  CC2931.L6_rep2 \n",
      "      1.5233362       1.4395133       1.0948225       1.5069215       1.6812947 \n",
      " CC2931.L7_rep1  CC2931.L7_rep3  CC2931.L9_rep1 CC2931.L13_rep1  CC2931.L4_rep3 \n",
      "      0.9991658       0.4242726       0.3583392       0.4939641       0.9496580 \n",
      " CC2931.L5_rep2  CC2931.L6_rep1  CC2931.L6_rep3  CC2931.L7_rep2 CC2931.L11_rep3 \n",
      "      1.4963887       1.2747691       1.8245148       1.2956589       0.7424569 \n",
      " CC2931.L9_rep2  CC2931.L9_rep3 CC2931.L13_rep2 CC2931.L10_rep1 CC2931.L13_rep3 \n",
      "      0.3790061       1.0412170       1.1207979       0.5070304       1.2204579 \n",
      "CC2931.L10_rep2 CC2931.L14_rep1 CC2931.L10_rep3 CC2931.L14_rep2 CC2931.L11_rep1 \n",
      "      0.6088216       1.6119599       0.5004204       0.9491339       1.7082821 \n",
      "CC2931.L14_rep3 CC2931.L11_rep2 \n",
      "      0.5156808       1.9069923 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "################\n",
    "#### CC2931 ####\n",
    "################\n",
    "CC2931_raw_counts <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2931_raw.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = \"geneid\")\n",
    "counts <- CC2931_raw_counts[c(1:42)]\n",
    "\n",
    "#### Assigning factors and conditions ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L9', 'L10', 'L11', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### Creating data object ####\n",
    "CC2931_dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "CC2931_dds <- estimateSizeFactors(CC2931_dds)\n",
    "sizeFactors(CC2931_dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "effective-combine",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CC2931.ANC.rep1 1.850294\n",
      "CC2931.ANC.rep2 1.0322654\n",
      "CC2931.ANC.rep3 0.8007557\n",
      "CC2931.L1.rep1 2.3773915\n",
      "CC2931.L1.rep2 1.3148481\n",
      "CC2931.L1.rep3 1.4059812\n",
      "CC2931.L2.rep1 0.5266999\n",
      "CC2931.L2.rep2 0.9162189\n",
      "CC2931.L2.rep3 1.0550166\n",
      "CC2931.L15_rep1 1.4258836\n",
      "CC2931.L15_rep2 1.0288307\n",
      "CC2931.L15_rep3 1.2657351\n",
      "CC2931.L3_rep1 0.6031843\n",
      "CC2931.L3_rep2 1.1181484\n",
      "CC2931.L3_rep3 0.8886967\n",
      "CC2931.L4_rep1 1.5233362\n",
      "CC2931.L4_rep2 1.4395133\n",
      "CC2931.L5_rep1 1.0948225\n",
      "CC2931.L5_rep3 1.5069215\n",
      "CC2931.L6_rep2 1.6812947\n",
      "CC2931.L7_rep1 0.9991658\n",
      "CC2931.L7_rep3 0.4242726\n",
      "CC2931.L9_rep1 0.3583392\n",
      "CC2931.L13_rep1 0.4939641\n",
      "CC2931.L4_rep3 0.949658\n",
      "CC2931.L5_rep2 1.4963887\n",
      "CC2931.L6_rep1 1.2747691\n",
      "CC2931.L6_rep3 1.8245148\n",
      "CC2931.L7_rep2 1.2956589\n",
      "CC2931.L11_rep3 0.7424569\n",
      "CC2931.L9_rep2 0.3790061\n",
      "CC2931.L9_rep3 1.041217\n",
      "CC2931.L13_rep2 1.1207979\n",
      "CC2931.L10_rep1 0.5070304\n",
      "CC2931.L13_rep3 1.2204579\n",
      "CC2931.L10_rep2 0.6088216\n",
      "CC2931.L14_rep1 1.6119599\n",
      "CC2931.L10_rep3 0.5004204\n",
      "CC2931.L14_rep2 0.9491339\n",
      "CC2931.L11_rep1 1.7082821\n",
      "CC2931.L14_rep3 0.5156808\n",
      "CC2931.L11_rep2 1.9069923\n"
     ]
    }
   ],
   "source": [
    "string = \"\"\"CC2931.ANC.rep1 CC2931.ANC.rep2 CC2931.ANC.rep3  CC2931.L1.rep1  CC2931.L1.rep2 \n",
    "      1.8502940       1.0322654       0.8007557       2.3773915       1.3148481 \n",
    " CC2931.L1.rep3  CC2931.L2.rep1  CC2931.L2.rep2  CC2931.L2.rep3 CC2931.L15_rep1 \n",
    "      1.4059812       0.5266999       0.9162189       1.0550166       1.4258836 \n",
    "CC2931.L15_rep2 CC2931.L15_rep3  CC2931.L3_rep1  CC2931.L3_rep2  CC2931.L3_rep3 \n",
    "      1.0288307       1.2657351       0.6031843       1.1181484       0.8886967 \n",
    " CC2931.L4_rep1  CC2931.L4_rep2  CC2931.L5_rep1  CC2931.L5_rep3  CC2931.L6_rep2 \n",
    "      1.5233362       1.4395133       1.0948225       1.5069215       1.6812947 \n",
    " CC2931.L7_rep1  CC2931.L7_rep3  CC2931.L9_rep1 CC2931.L13_rep1  CC2931.L4_rep3 \n",
    "      0.9991658       0.4242726       0.3583392       0.4939641       0.9496580 \n",
    " CC2931.L5_rep2  CC2931.L6_rep1  CC2931.L6_rep3  CC2931.L7_rep2 CC2931.L11_rep3 \n",
    "      1.4963887       1.2747691       1.8245148       1.2956589       0.7424569 \n",
    " CC2931.L9_rep2  CC2931.L9_rep3 CC2931.L13_rep2 CC2931.L10_rep1 CC2931.L13_rep3 \n",
    "      0.3790061       1.0412170       1.1207979       0.5070304       1.2204579 \n",
    "CC2931.L10_rep2 CC2931.L14_rep1 CC2931.L10_rep3 CC2931.L14_rep2 CC2931.L11_rep1 \n",
    "      0.6088216       1.6119599       0.5004204       0.9491339       1.7082821 \n",
    "CC2931.L14_rep3 CC2931.L11_rep2 \n",
    "      0.5156808       1.9069923\"\"\"\n",
    "names=[]\n",
    "sfs = []\n",
    "for i in string.strip().split():\n",
    "    if i[0] ==\"C\":names.append(i)\n",
    "    else: sfs.append(float(i))\n",
    "\n",
    "for i, j in zip(names, sfs):\n",
    "    print(i, j)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "elect-bishop",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CC2344.L15.rep3 CC2344.L15.rep2 CC2344.ANC.rep2  CC2344.L1.rep1  CC2344.L1.rep3 \n",
      "      1.1737230       1.3257651       1.3987882       1.8477988       1.9994374 \n",
      " CC2344.L2.rep2  CC2344.L3.rep1  CC2344.L3.rep3  CC2344.L4.rep2  CC2344.L5.rep1 \n",
      "      0.4575387       0.7316950       0.6404584       1.4682027       0.2274608 \n",
      " CC2344.L5.rep3  CC2344.L6.rep2  CC2344.L7.rep1  CC2344.L8.rep1  CC2344.L8.rep3 \n",
      "      0.7020427       0.4681475       0.5782607       0.9354567       0.7032777 \n",
      " CC2344.L9.rep2 CC2344.L10.rep1 CC2344.L10.rep3 CC2344.L11.rep2 CC2344.L12.rep1 \n",
      "      3.2269203       0.8035964       0.3408698       1.2464077       1.2621707 \n",
      "CC2344.L12.rep3 CC2344.L13.rep2 CC2344.L14.rep1 CC2344.L14.rep3  CC2344.L7.rep3 \n",
      "      0.5445667       1.4719460       1.4588763       1.1712411       0.5062961 \n",
      "CC2344.ANC.rep1 CC2344.ANC.rep3  CC2344.L1.rep2  CC2344.L2.rep1  CC2344.L2.rep3 \n",
      "      0.5721730       1.9936471       0.9835393       0.7979124       1.3583711 \n",
      " CC2344.L3.rep2  CC2344.L4.rep1  CC2344.L4.rep3  CC2344.L5.rep2  CC2344.L6.rep1 \n",
      "      1.8257090       0.9201224       0.7725692       1.2138964       0.7650153 \n",
      " CC2344.L6.rep3  CC2344.L7.rep2  CC2344.L8.rep2  CC2344.L9.rep1  CC2344.L9.rep3 \n",
      "      1.3146830       0.6315503       1.6609855       2.0764098       1.9049752 \n",
      "CC2344.L10.rep2 CC2344.L11.rep1 CC2344.L11.rep3 CC2344.L12.rep2 CC2344.L13.rep1 \n",
      "      0.9330056       1.6455449       1.2592621       1.1726861       0.8356573 \n",
      "CC2344.L13.rep3 CC2344.L14.rep2 CC2344.L15.rep1 \n",
      "      1.2341825       1.2795932       0.7815065 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "################\n",
    "#### CC2344 ####\n",
    "################\n",
    "CC2344_raw_counts <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2344_raw.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = 'geneid')\n",
    "counts <- CC2344_raw_counts[c(1:48)]\n",
    "\n",
    "#### Assigning factors and conditions ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L8', 'L9', 'L10', 'L11', 'L12', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### Creating data object ####\n",
    "CC2344_dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "CC2344_dds <- estimateSizeFactors(CC2344_dds)\n",
    "sizeFactors(CC2344_dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "spread-casino",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CC2344.L15.rep3 1.173723\n",
      "CC2344.L15.rep2 1.3257651\n",
      "CC2344.ANC.rep2 1.3987882\n",
      "CC2344.L1.rep1 1.8477988\n",
      "CC2344.L1.rep3 1.9994374\n",
      "CC2344.L2.rep2 0.4575387\n",
      "CC2344.L3.rep1 0.731695\n",
      "CC2344.L3.rep3 0.6404584\n",
      "CC2344.L4.rep2 1.4682027\n",
      "CC2344.L5.rep1 0.2274608\n",
      "CC2344.L5.rep3 0.7020427\n",
      "CC2344.L6.rep2 0.4681475\n",
      "CC2344.L7.rep1 0.5782607\n",
      "CC2344.L8.rep1 0.9354567\n",
      "CC2344.L8.rep3 0.7032777\n",
      "CC2344.L9.rep2 3.2269203\n",
      "CC2344.L10.rep1 0.8035964\n",
      "CC2344.L10.rep3 0.3408698\n",
      "CC2344.L11.rep2 1.2464077\n",
      "CC2344.L12.rep1 1.2621707\n",
      "CC2344.L12.rep3 0.5445667\n",
      "CC2344.L13.rep2 1.471946\n",
      "CC2344.L14.rep1 1.4588763\n",
      "CC2344.L14.rep3 1.1712411\n",
      "CC2344.L7.rep3 0.5062961\n",
      "CC2344.ANC.rep1 0.572173\n",
      "CC2344.ANC.rep3 1.9936471\n",
      "CC2344.L1.rep2 0.9835393\n",
      "CC2344.L2.rep1 0.7979124\n",
      "CC2344.L2.rep3 1.3583711\n",
      "CC2344.L3.rep2 1.825709\n",
      "CC2344.L4.rep1 0.9201224\n",
      "CC2344.L4.rep3 0.7725692\n",
      "CC2344.L5.rep2 1.2138964\n",
      "CC2344.L6.rep1 0.7650153\n",
      "CC2344.L6.rep3 1.314683\n",
      "CC2344.L7.rep2 0.6315503\n",
      "CC2344.L8.rep2 1.6609855\n",
      "CC2344.L9.rep1 2.0764098\n",
      "CC2344.L9.rep3 1.9049752\n",
      "CC2344.L10.rep2 0.9330056\n",
      "CC2344.L11.rep1 1.6455449\n",
      "CC2344.L11.rep3 1.2592621\n",
      "CC2344.L12.rep2 1.1726861\n",
      "CC2344.L13.rep1 0.8356573\n",
      "CC2344.L13.rep3 1.2341825\n",
      "CC2344.L14.rep2 1.2795932\n",
      "CC2344.L15.rep1 0.7815065\n"
     ]
    }
   ],
   "source": [
    "string = \"\"\"CC2344.L15.rep3 CC2344.L15.rep2 CC2344.ANC.rep2  CC2344.L1.rep1  CC2344.L1.rep3 \n",
    "      1.1737230       1.3257651       1.3987882       1.8477988       1.9994374 \n",
    " CC2344.L2.rep2  CC2344.L3.rep1  CC2344.L3.rep3  CC2344.L4.rep2  CC2344.L5.rep1 \n",
    "      0.4575387       0.7316950       0.6404584       1.4682027       0.2274608 \n",
    " CC2344.L5.rep3  CC2344.L6.rep2  CC2344.L7.rep1  CC2344.L8.rep1  CC2344.L8.rep3 \n",
    "      0.7020427       0.4681475       0.5782607       0.9354567       0.7032777 \n",
    " CC2344.L9.rep2 CC2344.L10.rep1 CC2344.L10.rep3 CC2344.L11.rep2 CC2344.L12.rep1 \n",
    "      3.2269203       0.8035964       0.3408698       1.2464077       1.2621707 \n",
    "CC2344.L12.rep3 CC2344.L13.rep2 CC2344.L14.rep1 CC2344.L14.rep3  CC2344.L7.rep3 \n",
    "      0.5445667       1.4719460       1.4588763       1.1712411       0.5062961 \n",
    "CC2344.ANC.rep1 CC2344.ANC.rep3  CC2344.L1.rep2  CC2344.L2.rep1  CC2344.L2.rep3 \n",
    "      0.5721730       1.9936471       0.9835393       0.7979124       1.3583711 \n",
    " CC2344.L3.rep2  CC2344.L4.rep1  CC2344.L4.rep3  CC2344.L5.rep2  CC2344.L6.rep1 \n",
    "      1.8257090       0.9201224       0.7725692       1.2138964       0.7650153 \n",
    " CC2344.L6.rep3  CC2344.L7.rep2  CC2344.L8.rep2  CC2344.L9.rep1  CC2344.L9.rep3 \n",
    "      1.3146830       0.6315503       1.6609855       2.0764098       1.9049752 \n",
    "CC2344.L10.rep2 CC2344.L11.rep1 CC2344.L11.rep3 CC2344.L12.rep2 CC2344.L13.rep1 \n",
    "      0.9330056       1.6455449       1.2592621       1.1726861       0.8356573 \n",
    "CC2344.L13.rep3 CC2344.L14.rep2 CC2344.L15.rep1 \n",
    "      1.2341825       1.2795932       0.7815065\"\"\"\n",
    "names=[]\n",
    "sfs = []\n",
    "for i in string.strip().split():\n",
    "    if i[0] ==\"C\":names.append(i)\n",
    "    else: sfs.append(float(i))\n",
    "\n",
    "for i, j in zip(names, sfs):\n",
    "    print(i, j)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "informational-sunrise",
   "metadata": {},
   "source": [
    "### DESeq2 normalizing raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "norwegian-insured",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "################\n",
    "#### CC2931 ####\n",
    "################\n",
    "CC2931_raw_counts <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2931_raw.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = \"geneid\")\n",
    "counts <- CC2931_raw_counts[c(1:42)]\n",
    "\n",
    "#### Assigning factors and conditions ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L9', 'L10', 'L11', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### Creating data object ####\n",
    "CC2931_dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "CC2931_dds <- estimateSizeFactors(CC2931_dds)\n",
    "counts <- counts(CC2931_dds, normalized=TRUE)\n",
    "write.table(counts, paste(directory, \"CC2931_deseq_normalized\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)\n",
    "\n",
    "################\n",
    "#### CC2344 ####\n",
    "################\n",
    "CC2344_raw_counts <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2344_raw.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = 'geneid')\n",
    "counts <- CC2344_raw_counts[c(1:48)]\n",
    "\n",
    "#### Assigning factors and conditions ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L8', 'L9', 'L10', 'L11', 'L12', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### Creating data object ####\n",
    "CC2344_dds <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)\n",
    "directory <- \"/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/\"\n",
    "CC2344_dds <- estimateSizeFactors(CC2344_dds)\n",
    "counts <- counts(CC2344_dds, normalized=TRUE)\n",
    "write.table(counts, paste(directory, \"CC2344_deseq_normalized\", \".txt\", sep=\"\"), sep=\"\\t\", quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affecting-difficulty",
   "metadata": {},
   "source": [
    "### Adding GC content to DESeq2 normalized gene counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "adapted-cuisine",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Opening GC length and fragment length\n",
    "gc_content = pd.read_csv('/research/projects/chlamydomonas/MAexpression/genome_info/v6_genome_plus_anno/GC_lengths.tsv', delimiter = '\\t')\n",
    "fragment_length = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/fpkm/fragment_length.csv', delimiter = '\\t')\n",
    "fragment_length = fragment_length.set_index('index')\n",
    "\n",
    "## Opening DESeq2 normalized gene counts\n",
    "CC2931_normalized_GC = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_deseq_normalized.txt', delimiter = '\\t')\n",
    "CC2344_normalized_GC = pd.read_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_deseq_normalized.txt', delimiter = '\\t')\n",
    "\n",
    "CC2931_normalized_GC = CC2931_normalized_GC.join([fragment_length, gc_content['GC']])\n",
    "CC2344_normalized_GC = CC2344_normalized_GC.join([fragment_length, gc_content['GC']])\n",
    "\n",
    "## Export ##\n",
    "CC2931_normalized_GC.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2931_deseq_normalized_wt_GC', sep = '\\t', index = True, header = True)\n",
    "CC2344_normalized_GC.to_csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/CC2344_deseq_normalized_wt_GC', sep = '\\t', index = True, header = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "healthy-holly",
   "metadata": {},
   "source": [
    "## DESeq2 - Differential Expression Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "third-protocol",
   "metadata": {},
   "source": [
    "### CC-2931"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "institutional-yacht",
   "metadata": {},
   "source": [
    "Creating DESeq2 data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "experienced-bobby",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "CC2931_raw_counts <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2931_raw.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = \"geneid\")\n",
    "counts <- CC2931_raw_counts[c(1:42)]\n",
    "\n",
    "#### Assigning factors and conditions ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L9', 'L10', 'L11', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### Creating data object ####\n",
    "CC2931_dds1 <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "danish-potential",
   "metadata": {},
   "source": [
    "Running DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "verified-police",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates: 4 workers\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates, fitting model and testing: 4 workers\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "#### USING DESEQ2 #####\n",
    "dds <- DESeq(CC2931_dds1, parallel = TRUE)\n",
    "res_L1 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L1'), alpha = 0.05)\n",
    "res_L2 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L2'), alpha = 0.05)\n",
    "res_L3 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L3'), alpha = 0.05)\n",
    "res_L4 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L4'), alpha = 0.05)\n",
    "res_L5 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L5'), alpha = 0.05)\n",
    "res_L6 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L6'), alpha = 0.05)\n",
    "res_L7 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L7'), alpha = 0.05)\n",
    "res_L9 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L9'), alpha = 0.05)\n",
    "res_L10 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L10'), alpha = 0.05)\n",
    "res_L11 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L11'), alpha = 0.05)\n",
    "res_L13 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L13'), alpha = 0.05)\n",
    "res_L14 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L14'), alpha = 0.05)\n",
    "res_L15 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L15'), alpha = 0.05)\n",
    "\n",
    "\n",
    "#### FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "resSig_L1 = res_L1[which(res_L1$padj < 0.05),]\n",
    "resSig_L2 = res_L2[which(res_L2$padj < 0.05),]\n",
    "resSig_L3 = res_L3[which(res_L3$padj < 0.05),]\n",
    "resSig_L4 = res_L4[which(res_L4$padj < 0.05),]\n",
    "resSig_L5 = res_L5[which(res_L5$padj < 0.05),]\n",
    "resSig_L6 = res_L6[which(res_L6$padj < 0.05),]\n",
    "resSig_L7 = res_L7[which(res_L7$padj < 0.05),]\n",
    "resSig_L9 = res_L9[which(res_L9$padj < 0.05),]\n",
    "resSig_L10 = res_L10[which(res_L10$padj < 0.05),]\n",
    "resSig_L11 = res_L11[which(res_L11$padj < 0.05),]\n",
    "resSig_L13 = res_L13[which(res_L13$padj < 0.05),]\n",
    "resSig_L14 = res_L14[which(res_L14$padj < 0.05),]\n",
    "resSig_L15 = res_L15[which(res_L15$padj < 0.05),]\n",
    "\n",
    "#### EXPORTING FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "write.csv(as.data.frame(resSig_L1), file=\"DEGs/degs/CC2931.resSig_L1.genes\")\n",
    "write.csv(as.data.frame(resSig_L2), file=\"DEGs/degs/CC2931.resSig_L2.genes\")\n",
    "write.csv(as.data.frame(resSig_L3), file=\"DEGs/degs/CC2931.resSig_L3.genes\")\n",
    "write.csv(as.data.frame(resSig_L4), file=\"DEGs/degs/CC2931.resSig_L4.genes\")\n",
    "write.csv(as.data.frame(resSig_L5), file=\"DEGs/degs/CC2931.resSig_L5.genes\")\n",
    "write.csv(as.data.frame(resSig_L6), file=\"DEGs/degs/CC2931.resSig_L6.genes\")\n",
    "write.csv(as.data.frame(resSig_L7), file=\"DEGs/degs/CC2931.resSig_L7.genes\")\n",
    "write.csv(as.data.frame(resSig_L9), file=\"DEGs/degs/CC2931.resSig_L9.genes\")\n",
    "write.csv(as.data.frame(resSig_L10), file=\"DEGs/degs/CC2931.resSig_L10.genes\")\n",
    "write.csv(as.data.frame(resSig_L11), file=\"DEGs/degs/CC2931.resSig_L11.genes\")\n",
    "write.csv(as.data.frame(resSig_L13), file=\"DEGs/degs/CC2931.resSig_L13.genes\")\n",
    "write.csv(as.data.frame(resSig_L14), file=\"DEGs/degs/CC2931.resSig_L14.genes\")\n",
    "write.csv(as.data.frame(resSig_L15), file=\"DEGs/degs/CC2931.resSig_L15.genes\")\n",
    "\n",
    "#### EXPORTING LOG2FOLD CHANGE RESULTS ####\n",
    "write.table(as.data.frame(res_L1), 'genes_log2fold/CC2931-L1',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L2), 'genes_log2fold/CC2931-L2',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L3), 'genes_log2fold/CC2931-L3',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L4), 'genes_log2fold/CC2931-L4',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L5), 'genes_log2fold/CC2931-L5',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L6), 'genes_log2fold/CC2931-L6',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L7), 'genes_log2fold/CC2931-L7',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L9), 'genes_log2fold/CC2931-L9',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L10), 'genes_log2fold/CC2931-L10',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L11), 'genes_log2fold/CC2931-L11',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L13), 'genes_log2fold/CC2931-L13',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L14), 'genes_log2fold/CC2931-L14',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L15), 'genes_log2fold/CC2931-L15',  sep=\"\\t\", quote = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "disabled-england",
   "metadata": {},
   "source": [
    "### CC-2344"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "friendly-mixer",
   "metadata": {},
   "source": [
    "Creating DESeq2 data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ahead-simple",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "CC2344_raw_counts <- read.csv('/research/projects/chlamydomonas/MAexpression/analysis/raw_counts/unnormalized_CC2344_raw.txt', sep = '\\t', header = TRUE, stringsAsFactors = FALSE, row.names = 'geneid')\n",
    "counts <- CC2344_raw_counts[c(1:48)]\n",
    "\n",
    "#### Assigning factors and conditions ####\n",
    "condition <- c('ANC', 'L2', 'L3', 'L4', 'L5', 'L6', 'L7', 'L8', 'L9', 'L10', 'L11', 'L12', 'L13', 'L14', 'L15')\n",
    "factor <- colnames(counts)\n",
    "columns <- factor\n",
    "for (i in condition) {factor[grep((i), factor)] <- i}\n",
    "factor[grep('L1.rep', factor)] <- 'L1'\n",
    "coldata <- data.frame(factor, columns, row.names = 'columns')\n",
    "\n",
    "#### Creating data object ####\n",
    "CC2344_dds1 <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~ factor)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "juvenile-ticket",
   "metadata": {},
   "source": [
    "Running DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "advance-football",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates: 4 workers\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "#### USING DESEQ2 #####\n",
    "dds <- DESeq(CC2344_dds1, parallel = TRUE)\n",
    "res_L1 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L1'), alpha = 0.05)\n",
    "res_L2 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L2'), alpha = 0.05)\n",
    "res_L3 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L3'), alpha = 0.05)\n",
    "res_L4 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L4'), alpha = 0.05)\n",
    "res_L5 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L5'), alpha = 0.05)\n",
    "res_L6 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L6'), alpha = 0.05)\n",
    "res_L7 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L7'), alpha = 0.05)\n",
    "res_L8 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L8'), alpha = 0.05)\n",
    "res_L9 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L9'), alpha = 0.05)\n",
    "res_L10 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L10'), alpha = 0.05)\n",
    "res_L11 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L11'), alpha = 0.05)\n",
    "res_L12 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L12'), alpha = 0.05)\n",
    "res_L13 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L13'), alpha = 0.05)\n",
    "res_L14 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L14'), alpha = 0.05)\n",
    "res_L15 <- results(dds, pAdjustMethod = \"BH\", contrast = c('factor', 'ANC','L15'), alpha = 0.05)\n",
    "\n",
    "#### FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "resSig_L1 = res_L1[which(res_L1$padj < 0.05),]\n",
    "resSig_L2 = res_L2[which(res_L2$padj < 0.05),]\n",
    "resSig_L3 = res_L3[which(res_L3$padj < 0.05),]\n",
    "resSig_L4 = res_L4[which(res_L4$padj < 0.05),]\n",
    "resSig_L5 = res_L5[which(res_L5$padj < 0.05),]\n",
    "resSig_L6 = res_L6[which(res_L6$padj < 0.05),]\n",
    "resSig_L7 = res_L7[which(res_L7$padj < 0.05),]\n",
    "resSig_L8 = res_L8[which(res_L8$padj < 0.05),]\n",
    "resSig_L9 = res_L9[which(res_L9$padj < 0.05),]\n",
    "resSig_L10 = res_L10[which(res_L10$padj < 0.05),]\n",
    "resSig_L11 = res_L11[which(res_L11$padj < 0.05),]\n",
    "resSig_L12 = res_L12[which(res_L12$padj < 0.05),]\n",
    "resSig_L13 = res_L13[which(res_L13$padj < 0.05),]\n",
    "resSig_L14 = res_L14[which(res_L14$padj < 0.05),]\n",
    "resSig_L15 = res_L15[which(res_L15$padj < 0.05),]\n",
    "\n",
    "#### EXPORTING FILES OF DIFFERNTIALLY EXPRESSED GENES ####\n",
    "write.csv(as.data.frame(resSig_L1), file=\"DEGs/degs/CC2344.resSig_L1.genes\")\n",
    "write.csv(as.data.frame(resSig_L2), file=\"DEGs/degs/CC2344.resSig_L2.genes\")\n",
    "write.csv(as.data.frame(resSig_L3), file=\"DEGs/degs/CC2344.resSig_L3.genes\")\n",
    "write.csv(as.data.frame(resSig_L4), file=\"DEGs/degs/CC2344.resSig_L4.genes\")\n",
    "write.csv(as.data.frame(resSig_L5), file=\"DEGs/degs/CC2344.resSig_L5.genes\")\n",
    "write.csv(as.data.frame(resSig_L6), file=\"DEGs/degs/CC2344.resSig_L6.genes\")\n",
    "write.csv(as.data.frame(resSig_L7), file=\"DEGs/degs/CC2344.resSig_L7.genes\")\n",
    "write.csv(as.data.frame(resSig_L8), file=\"DEGs/degs/CC2344.resSig_L8.genes\")\n",
    "write.csv(as.data.frame(resSig_L9), file=\"DEGs/degs/CC2344.resSig_L9.genes\")\n",
    "write.csv(as.data.frame(resSig_L10), file=\"DEGs/degs/CC2344.resSig_L10.genes\")\n",
    "write.csv(as.data.frame(resSig_L11), file=\"DEGs/degs/CC2344.resSig_L11.genes\")\n",
    "write.csv(as.data.frame(resSig_L12), file=\"DEGs/degs/CC2344.resSig_L12.genes\")\n",
    "write.csv(as.data.frame(resSig_L13), file=\"DEGs/degs/CC2344.resSig_L13.genes\")\n",
    "write.csv(as.data.frame(resSig_L14), file=\"DEGs/degs/CC2344.resSig_L14.genes\")\n",
    "write.csv(as.data.frame(resSig_L15), file=\"DEGs/degs/CC2344.resSig_L15.genes\")\n",
    "\n",
    "#### EXPORTING LOG2FOLD CHANGE RESULTS ####\n",
    "write.table(as.data.frame(res_L1), 'genes_log2fold/CC2344-L1',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L2), 'genes_log2fold/CC2344-L2',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L3), 'genes_log2fold/CC2344-L3',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L4), 'genes_log2fold/CC2344-L4',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L5), 'genes_log2fold/CC2344-L5',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L6), 'genes_log2fold/CC2344-L6',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L7), 'genes_log2fold/CC2344-L7',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L8), 'genes_log2fold/CC2344-L8',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L9), 'genes_log2fold/CC2344-L9',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L10), 'genes_log2fold/CC2344-L10',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L11), 'genes_log2fold/CC2344-L11',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L12), 'genes_log2fold/CC2344-L12',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L13), 'genes_log2fold/CC2344-L13',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L14), 'genes_log2fold/CC2344-L14',  sep=\"\\t\", quote = F)\n",
    "write.table(as.data.frame(res_L15), 'genes_log2fold/CC2344-L15',  sep=\"\\t\", quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "naked-queen",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-personal] *",
   "language": "python",
   "name": "conda-env-.conda-personal-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
